### PR TITLE
S2FRMS-40/S2FRMS-58 child select

### PR DESCRIPTION
**Pull request by stanbellcom**
S2FRMS-40 child select

### CODE REVIEW

File: modules/os2forms_nemid/src/Plugin/WebformElement/NemidChildrenSelect.php [hunk 1]
```diff
@@ -136,23 +134,30 @@ public function handleElementPrepopulate(array &$element, FormStateInterface &$f
...
             }
           }
         }
       }
     }
 
     $element['#options'] = $options;
+
+    if ($showAddressNameProtectionMessage) {
+      $element['#suffix'] = $element['#address_protection_help_text'];
```
Comment (rimi-itk) [inline modules/os2forms_nemid/src/Plugin/WebformElement/NemidChildrenSelect.php:159]: It may be a good idea to wrap the help text in a `p` or `div` element.

File: modules/os2forms_nemid/src/Plugin/WebformElement/NemidChildrenSelect.php [hunk 2]
```diff
@@ -136,23 +134,30 @@ public function handleElementPrepopulate(array &$element, FormStateInterface &$f
...
+
     if ($cprLookupResult) {
       $prepopulateKey = $this->getPrepopulateFieldFieldKey($element);
       if ($children = $cprLookupResult->getFieldValue($prepopulateKey)) {
         if (is_array($children) && !empty($children)) {
           foreach ($children as $child) {
-            if ($element['#cpr_output_type'] == 'cpr') {
-              $options[$child['cpr']] = $child['name'];
+            if ($child['nameAddressProtected']) {
+              $options[$child['cpr']] = $child['cpr'] . ' (Navne- og adressebeskyttet)';
```
Comment (rimi-itk) [inline modules/os2forms_nemid/src/Plugin/WebformElement/NemidChildrenSelect.php:145]: “(Navne- og adressebeskyttet)” should be translatable (OS2Forms support multi-lingual forms).

File: modules/os2forms_nemid/src/Plugin/WebformElement/ServiceplatformenCprElementBase.php [hunk 3]
```diff
@@ -27,8 +28,26 @@ public function handleElementPrepopulate(array &$element, FormStateInterface &$f
...
+   * Appends name/address protected text, if person has name/address protection.
+   *
+   * @param \Drupal\os2web_datalookup\LookupResult\CprLookupResult $cprLookupResult
+   *   Initialized CprLooupResult.
+   *
+   * @return string
+   *   String " (Navne- og adressebeskyttet)" or nothing.
+   */
+  public function appendNameAddressProtectedText(CprLookupResult $cprLookupResult) {
+    return $cprLookupResult->isNameAddressProtected() ? ' (Navne- og adressebeskyttet)' : '';
```
Comment (rimi-itk) [inline modules/os2forms_nemid/src/Plugin/WebformElement/ServiceplatformenCprElementBase.php:50]: “(Navne- og adressebeskyttet)” should be translatable (OS2Forms support multi-lingual forms).

File: modules/os2forms_nemid/src/Plugin/WebformElement/ServiceplatformenCprElementBase.php [hunk 4]
```diff
@@ -27,8 +28,26 @@ public function handleElementPrepopulate(array &$element, FormStateInterface &$f
...
+  /**
+   * Appends name/address protected text, if person has name/address protection.
+   *
+   * @param \Drupal\os2web_datalookup\LookupResult\CprLookupResult $cprLookupResult
+   *   Initialized CprLooupResult.
+   *
+   * @return string
+   *   String " (Navne- og adressebeskyttet)" or nothing.
+   */
+  public function appendNameAddressProtectedText(CprLookupResult $cprLookupResult) {
```
Comment (rimi-itk) [inline modules/os2forms_nemid/src/Plugin/WebformElement/ServiceplatformenCprElementBase.php:49]: This function does not append anything, but returns something that should be appended by the caller. Rename to `getNameAddressProtectionText`, say, or inline the function if it's only used once.

File: modules/os2forms_nemid/src/Service/FormsHelper.php [hunk 5]
```diff
@@ -129,6 +129,27 @@ public function retrieveCprLookupResult(FormStateInterface $form_state) {
...
+  public static function retrieveCachedCprLookupResult(FormStateInterface $form_state) {
+    /** @var \Drupal\os2web_datalookup\LookupResult\CprLookupResult $cprLookupResult */
+    $cprLookupResult = NULL;
+
+    // Trying to fetch person data from cache.
+    if ($form_state->has('cprLookupResult')) {
+      $cprLookupResult = $form_state->get('cprLookupResult');
+    }
+
+    return $cprLookupResult;
```
Comment (rimi-itk) [inline modules/os2forms_nemid/src/Service/FormsHelper.php:150]:
```suggestion
    return $form_state->has('cprLookupResult') ? $form_state->get('cprLookupResult') : NULL;
```

**Review by rimi-itk — CHANGES_REQUESTED** [review]
Looks good. Some minor changes requested.

**Comment by stanbellcom** [comment]
@rimi-itk added fixes, please check again

**Review by rimi-itk — APPROVED** [review]

File: modules/os2forms_nemid/src/Element/MitidChildrenSelectAjaxBehaviour.php [hunk 6]
```diff
@@ -59,7 +59,7 @@ public static function mitidChildrenSelectAjax(array &$form, FormStateInterface
 
           switch ($element['#type']) {
             case 'os2forms_mitid_child_name':
-              $element['#value'] = !$cprLookupResult->isNameAddressProtected() ? $cprLookupResult->getName() : 'Navne- og adressebeskyttet';
+              $element['#value'] = !$cprLookupResult->isNameAddressProtected() ? $cprLookupResult->getName() : t('Name and address protected');
```
Comment (rimi-itk) [inline modules/os2forms_nemid/src/Element/MitidChildrenSelectAjaxBehaviour.php:62]: You should use `StringTranslationTrait` and `$this->t` (rather than `t`) if possible.

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: modules/os2forms_nemid/src/Element/MitidChildrenSelectAjaxBehaviour.php (hunk 6 above)
Comment (stanbellcom) [reply modules/os2forms_nemid/src/Element/MitidChildrenSelectAjaxBehaviour.php:62]: @rimi-itk 
generally agree, but I could not change it in the specified file - it's used as static function, therefore using $this is impossible.
Changed it in other places/files where possible.

please check again

File: CHANGELOG.md [hunk 7]
```diff
@@ -13,6 +13,17 @@ before starting to add changes. Use example [placed in the end of the page](#exa
...
+
+- [#117](https://github.com/OS2Forms/os2forms/pull/117)
+  Encrypts all elements if encryption enabled.
+- [#114](https://github.com/OS2Forms/os2forms/pull/114)
+  Encrypted computed elements.
+
+## [3.15.3] 2024-06-25
+
+- [OS-74] Replacing DAWA matrikula select with Datafordeler select
+
```
Comment (rimi-itk) [inline CHANGELOG.md:26]: These changes do not seem related to this pull request.

File: modules/os2forms_nemid/src/Element/MitidChildrenSelectAjaxBehaviour.php (hunk 6 above)
Comment (rimi-itk) [reply modules/os2forms_nemid/src/Element/MitidChildrenSelectAjaxBehaviour.php:62]: Makes sense, @stanbellcom. One final comment: https://github.com/OS2Forms/os2forms/pull/111#discussion_r1673816493

File: CHANGELOG.md (hunk 7 above)
Comment (stanbellcom) [reply CHANGELOG.md:26]: these changes are not visible when you look at the pull request changes: into develop from f/S2FRMS-40_child_select:
https://github.com/OS2Forms/os2forms/pull/111/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed

It is in fact a resolution of the conflict we had in CHANGELOG.md file between two branches. If this conflict would have been resolved otherwise (meaning deleting these 10 lines), then upon merging these 10 lines would also be deleted from develop branch.

Comment (rimi-itk) [reply CHANGELOG.md:26]: Oops, my bad. Thanks for the explanation.

**Review by rimi-itk — APPROVED** [review]

**Comment by rimi-itk** [comment]
@stanbellcom, I've approved the changes. Good job!
